diff -upr -X dontdiff rtems/c/src/lib/libbsp/sparc64/usiii/make/custom/usiii.cfg rtems-hwsched2/c/src/lib/libbsp/sparc64/usiii/make/custom/usiii.cfg
--- rtems/c/src/lib/libbsp/sparc64/usiii/make/custom/usiii.cfg	2010-06-17 12:20:39.000000000 -0400
+++ rtems-hwsched2/c/src/lib/libbsp/sparc64/usiii/make/custom/usiii.cfg	2010-10-09 19:09:23.000000000 -0400
@@ -24,4 +24,4 @@ CPU_CFLAGS = -mcpu=ultrasparc3 -DUS3 -DS
 #CPU_CFLAGS = -mcpu=ultrasparc3 -DRTEMS_DEBUG -DHAVE___ASSERT_FUNC 
 
 # optimize flag: typically -O2
-CFLAGS_OPTIMIZE_V = -gstabs
+CFLAGS_OPTIMIZE_V = -gstabs -O2
diff -upr -X dontdiff rtems/cpukit/rtems/src/clocktick.c rtems-hwsched2/cpukit/rtems/src/clocktick.c
--- rtems/cpukit/rtems/src/clocktick.c	2010-07-08 13:33:01.000000000 -0400
+++ rtems-hwsched2/cpukit/rtems/src/clocktick.c	2010-10-09 18:23:29.000000000 -0400
@@ -41,12 +41,20 @@
 
 rtems_status_code rtems_clock_tick( void )
 {
+  ISR_Level          level;
+  asm volatile("flushw");
+  _ISR_Disable( level );
+  MAGIC_HWSCHED_START;
+
   _TOD_Tickle_ticks();
 
   _Watchdog_Tickle_ticks();
 
   _Thread_Tickle_timeslice();
 
+  MAGIC_HWSCHED_END;
+  _ISR_Enable( level );
+
   if ( _Thread_Is_context_switch_necessary() &&
        _Thread_Is_dispatching_enabled() )
     _Thread_Dispatch();
diff -upr -X dontdiff rtems/cpukit/score/cpu/sparc64/rtems/score/sparc64.h rtems-hwsched2/cpukit/score/cpu/sparc64/rtems/score/sparc64.h
--- rtems/cpukit/score/cpu/sparc64/rtems/score/sparc64.h	2010-10-09 17:38:15.000000000 -0400
+++ rtems-hwsched2/cpukit/score/cpu/sparc64/rtems/score/sparc64.h	2010-10-09 18:22:47.000000000 -0400
@@ -145,6 +145,66 @@ extern "C" {
 
 #define STACK_BIAS (2047)
 
+/* Additions to support MAGIC Simics commands */
+#ifdef ASM
+#define MAGIC(n) \
+  sethi n, %g0
+
+#endif /* ASM */
+
+#ifndef ASM
+
+#define __MAGIC_CASSERT(p) do {                                 \
+         typedef int __check_magic_argument[(p) ? 1 : -1];       \
+} while (0)
+
+#define MAGIC(n) do {                                   \
+    __MAGIC_CASSERT((n) > 0 && (n) < (1U << 22));   \
+         __asm__ __volatile__ ("sethi " #n ", %g0");     \
+} while (0)
+
+#define MAGIC_FLUSH(n) do { \
+    __MAGIC_CASSERT((n) > 0 && (n) < (1U << 22));   \
+        __asm__ __volatile__ ("flushw"); \
+         __asm__ __volatile__ ("sethi " #n ", %g0");     \
+} while (0)
+
+
+#endif /* !ASM */
+#define MAGIC_BREAKPOINT MAGIC(0x40000)
+
+#define MAGIC_HWSCHED_START     MAGIC(0x10)
+#define MAGIC_HWSCHED_UNPAUSE   MAGIC(0x20)
+
+#define MAGIC_HWSCHED_END       MAGIC(0x11)
+#define MAGIC_HWSCHED_PAUSE     MAGIC(0x21)
+
+
+#define MAGIC_HWDS1_START               MAGIC(0x100)
+#define MAGIC_HWDS1_FIRST               MAGIC(0x101)
+#define MAGIC_HWDS1_ENQUEUE             MAGIC(0x102)
+#define MAGIC_HWDS1_ENQUEUE_FIRST       MAGIC(0x103)
+#define MAGIC_HWDS1_EXTRACT             MAGIC(0x104)
+#define MAGIC_HWDS1_REQUEUE             MAGIC(0x105)
+
+#define MAGIC_HWDS1_END                 MAGIC(0x110)
+#define MAGIC_HWDS1_FIRST_END           MAGIC(0x111)
+#define MAGIC_HWDS1_ENQUEUE_END         MAGIC(0x112)
+#define MAGIC_HWDS1_ENQUEUE_FIRST_END   MAGIC(0x113)
+#define MAGIC_HWDS1_EXTRACT_END         MAGIC(0x114)
+#define MAGIC_HWDS1_REQUEUE_END         MAGIC(0x115)
+
+#define MAGIC_HWDS2_START               MAGIC(0x200)
+#define MAGIC_HWDS2_ENQUEUE             MAGIC(0x201)
+#define MAGIC_HWDS2_EXTRACT             MAGIC(0x202)
+#define MAGIC_HWDS2_UNPAUSE             MAGIC(0x400)
+
+#define MAGIC_HWDS2_END                 MAGIC(0x210)
+#define MAGIC_HWDS2_PAUSE               MAGIC(0x410)
+
+#ifdef ASM
+
+
 /* 
  * To enable the FPU we need to set both PSTATE.pef and FPRS.fef
  */
diff -upr -X dontdiff rtems/cpukit/score/include/rtems/score/isr.h rtems-hwsched2/cpukit/score/include/rtems/score/isr.h
--- rtems/cpukit/score/include/rtems/score/isr.h	2010-07-12 19:18:12.000000000 -0400
+++ rtems-hwsched2/cpukit/score/include/rtems/score/isr.h	2010-10-05 13:00:49.000000000 -0400
@@ -102,6 +102,7 @@ void _ISR_Handler_initialization ( void 
   do { \
     _CPU_ISR_Disable( _level ); \
     RTEMS_COMPILER_MEMORY_BARRIER(); \
+    MAGIC_HWSCHED_UNPAUSE; \
   } while (0)
 
 /**
@@ -111,6 +112,7 @@ void _ISR_Handler_initialization ( void 
  */
 #define _ISR_Enable( _level ) \
   do { \
+    MAGIC_HWSCHED_PAUSE; \
     RTEMS_COMPILER_MEMORY_BARRIER(); \
     _CPU_ISR_Enable( _level ); \
   } while (0)
diff -upr -X dontdiff rtems/cpukit/score/include/rtems/score/rbtree.h rtems-hwsched2/cpukit/score/include/rtems/score/rbtree.h
--- rtems/cpukit/score/include/rtems/score/rbtree.h	2010-07-20 13:06:23.000000000 -0400
+++ rtems-hwsched2/cpukit/score/include/rtems/score/rbtree.h	2010-10-09 19:09:23.000000000 -0400
@@ -245,11 +245,11 @@ extern "C" {
    *  @note It does NOT disable interrupts to ensure the atomicity
    *        of the extract operation.
    */
-  RBTree_Node *_RBTree_Insert_unprotected(
+/*  RBTree_Node *_RBTree_Insert_unprotected(
       RBTree_Control *the_rbtree,
       RBTree_Node *the_node
   );
-
+*/
   /**
    *  @brief Insert a node on a rbtree
    *
@@ -272,11 +272,11 @@ extern "C" {
    *        of the extract operation.
    */
 
-  void _RBTree_Extract_unprotected(
+/*  void _RBTree_Extract_unprotected(
       RBTree_Control *the_rbtree,
       RBTree_Node *the_node
   );
-
+*/
 
   /**
    *  @brief Delete a node from the rbtree
diff -upr -X dontdiff rtems/cpukit/score/include/rtems/score/scheduler.h rtems-hwsched2/cpukit/score/include/rtems/score/scheduler.h
--- rtems/cpukit/score/include/rtems/score/scheduler.h	2010-07-29 11:09:54.000000000 -0400
+++ rtems-hwsched2/cpukit/score/include/rtems/score/scheduler.h	2010-10-09 19:09:23.000000000 -0400
@@ -98,8 +98,6 @@ typedef struct {
   /** This field contains the thread's deadline information. */
   RBTree_Node                       deadline;
 
-  unsigned int                      absolute_deadline;
-
   /** 
    * This field points to the last node in the ready queue that has 
    * the same deadline (absolute) as this thread.
diff -upr -X dontdiff rtems/cpukit/score/inline/rtems/score/rbtree.inl rtems-hwsched2/cpukit/score/inline/rtems/score/rbtree.inl
--- rtems/cpukit/score/inline/rtems/score/rbtree.inl	2010-07-20 13:41:06.000000000 -0400
+++ rtems-hwsched2/cpukit/score/inline/rtems/score/rbtree.inl	2010-10-09 19:09:23.000000000 -0400
@@ -367,28 +367,6 @@ RTEMS_INLINE_ROUTINE RBTree_Node *_RBTre
   return iter_node;
 }
 
-/** @brief Get the First Node (unprotected)
- *
- *  This function removes the minimum or maximum node from the_rbtree and 
- *  returns a pointer to that node.  It does NOT disable interrupts to ensure
- *  the atomicity of the get operation.
- *
- *  @param[in] the_rbtree is the rbtree to attempt to get the min node from.
- *  @param[in] dir specifies whether to get minimum (0) or maximum (1)
- *
- *  @return This method returns the min or max node on the rbtree, or NULL.
- *
- *  @note This routine may return NULL if the RBTree is empty.
- */
-RTEMS_INLINE_ROUTINE RBTree_Node *_RBTree_Get_unprotected(
-    RBTree_Control *the_rbtree,
-    RBTree_Direction dir
-    )
-{
-  RBTree_Node *the_node = the_rbtree->first[dir];
-  _RBTree_Extract_unprotected(the_rbtree, the_node);
-  return the_node;
-}
 
 /** @brief Peek at the First Node (unprotected)
  *
@@ -436,5 +414,328 @@ RTEMS_INLINE_ROUTINE void _RBTree_Rotate
 }
 /**@}*/
 
+/** @brief Validate and fix-up tree properties for a new insert/colored node
+ *
+ *  This routine checks and fixes the Red-Black Tree properties based on 
+ *  @a the_node being just added to the tree.
+ * 
+ *  @note It does NOT disable interrupts to ensure the atomicity of the
+ *        append operation.
+ */
+RTEMS_INLINE_ROUTINE void _RBTree_Validate_insert_unprotected(
+    RBTree_Node    *the_node
+    )
+{
+  RBTree_Node *u,*g;
+
+  /* note: the insert root case is handled already */
+  /* if the parent is black, nothing needs to be done
+   * otherwise may need to loop a few times */
+  while (_RBTree_Is_red(_RBTree_Parent(the_node))) {
+    u = _RBTree_Parent_sibling(the_node);
+    g = the_node->parent->parent;
+    
+    /* if uncle is red, repaint uncle/parent black and grandparent red */
+    if(_RBTree_Is_red(u)) {
+      the_node->parent->color = RBT_BLACK;
+      u->color = RBT_BLACK;
+      g->color = RBT_RED;
+      the_node = g;
+    } else { /* if uncle is black */
+      RBTree_Direction dir = the_node != the_node->parent->child[0];
+      RBTree_Direction pdir = the_node->parent != g->child[0];
+      
+      /* ensure node is on the same branch direction as parent */
+      if (dir != pdir) {
+        _RBTree_Rotate(the_node->parent, pdir);
+        the_node = the_node->child[pdir];
+      }
+      the_node->parent->color = RBT_BLACK;
+      g->color = RBT_RED;
+      
+      /* now rotate grandparent in the other branch direction (toward uncle) */
+      _RBTree_Rotate(g, (1-pdir));
+    }
+  }
+  if(!the_node->parent->parent) the_node->color = RBT_BLACK;
+}
+
+/** @brief Insert a Node (unprotected)
+ *
+ *  This routine inserts @a the_node on the Red-Black Tree @a the_rbtree.
+ *
+ *  @retval 0 Successfully inserted.
+ *  @retval -1 NULL @a the_node.
+ *  @retval RBTree_Node* if one with equal value to @a the_node->value exists 
+ *          in @a the_rbtree.
+ *
+ *  @note It does NOT disable interrupts to ensure the atomicity
+ *        of the extract operation.
+ */
+RTEMS_INLINE_ROUTINE  RBTree_Node *_RBTree_Insert_unprotected(
+    RBTree_Control *the_rbtree,
+    RBTree_Node *the_node
+    )
+{
+  if(!the_node) return (RBTree_Node*)-1;
+
+  RBTree_Node *iter_node = the_rbtree->root;
+
+  if (!iter_node) { /* special case: first node inserted */
+    the_node->color = RBT_BLACK;
+    the_rbtree->root = the_node;
+    the_rbtree->first[0] = the_rbtree->first[1] = the_node;
+    the_node->parent = (RBTree_Node *) the_rbtree;
+    the_node->child[RBT_LEFT] = the_node->child[RBT_RIGHT] = NULL;
+  } else { 
+    /* typical binary search tree insert, descend tree to leaf and insert */
+    while (iter_node) {
+      if(the_node->value == iter_node->value) return(iter_node);
+      RBTree_Direction dir = the_node->value > iter_node->value;
+      if (!iter_node->child[dir]) {
+        the_node->child[RBT_LEFT] = the_node->child[RBT_RIGHT] = NULL;
+        the_node->color = RBT_RED;
+        iter_node->child[dir] = the_node;
+        the_node->parent = iter_node;
+        /* update min/max */
+        if (_RBTree_Is_first(the_rbtree, iter_node, dir)) {
+          the_rbtree->first[dir] = the_node;
+        }
+        break;
+      } else {
+        iter_node = iter_node->child[dir];
+      }
+
+    } /* while(iter_node) */
+    
+    /* verify red-black properties */
+    _RBTree_Validate_insert_unprotected(the_node);
+  }
+  return (RBTree_Node*)0;
+}
+
+/** @brief  Validate and fix-up tree properties after deleting a node
+ *
+ *  This routine is called on a black node, @a the_node, after its deletion.
+ *  This function maintains the properties of the red-black tree.
+ *
+ *  @note It does NOT disable interrupts to ensure the atomicity
+ *        of the extract operation.
+ */
+RTEMS_INLINE_ROUTINE void _RBTree_Extract_validate_unprotected(
+    RBTree_Node *the_node
+    )
+{
+  RBTree_Node *parent, *sibling;
+  RBTree_Direction dir;
+
+  parent = the_node->parent; 
+  if(!parent->parent) return;
+
+  sibling = _RBTree_Sibling(the_node);
+
+  /* continue to correct tree as long as the_node is black and not the root */
+  while (!_RBTree_Is_red(the_node) && parent->parent) {
+
+    /* if sibling is red, switch parent (black) and sibling colors, 
+     * then rotate parent left, making the sibling be the_node's grandparent.
+     * Now the_node has a black sibling and red parent. After rotation, 
+     * update sibling pointer.
+     */
+    if (_RBTree_Is_red(sibling)) {
+      parent->color = RBT_RED;
+      sibling->color = RBT_BLACK;
+      dir = the_node != parent->child[0];
+      _RBTree_Rotate(parent, dir);
+      sibling = parent->child[!dir];
+    }
+
+    /* sibling is black, see if both of its children are also black. */
+    if (sibling && 
+        !_RBTree_Is_red(sibling->child[RBT_RIGHT]) && 
+        !_RBTree_Is_red(sibling->child[RBT_LEFT])) {
+        sibling->color = RBT_RED;
+        if (_RBTree_Is_red(parent)) {
+          parent->color = RBT_BLACK;
+          break;
+        }
+        the_node = parent; /* done if parent is red */
+        parent = the_node->parent;
+        sibling = _RBTree_Sibling(the_node);
+    } else if(sibling) {
+      /* at least one of sibling's children is red. we now proceed in two 
+       * cases, either the_node is to the left or the right of the parent. 
+       * In both cases, first check if one of sibling's children is black, 
+       * and if so rotate in the proper direction and update sibling pointer.
+       * Then switch the sibling and parent colors, and rotate through parent.
+       */
+      dir = the_node != parent->child[0];
+      if (!_RBTree_Is_red(sibling->child[!dir])) {
+        sibling->color = RBT_RED;
+        sibling->child[dir]->color = RBT_BLACK;
+        _RBTree_Rotate(sibling, !dir);
+        sibling = parent->child[!dir];
+      }
+      sibling->color = parent->color;
+      parent->color = RBT_BLACK;
+      sibling->child[!dir]->color = RBT_BLACK;
+      _RBTree_Rotate(parent, dir);
+      break; /* done */
+    }
+  } /* while */
+  if(!the_node->parent->parent) the_node->color = RBT_BLACK;
+}
+
+/** @brief Extract a Node (unprotected)
+ *
+ *  This routine extracts (removes) @a the_node from @a the_rbtree.
+ *
+ *  @note It does NOT disable interrupts to ensure the atomicity
+ *        of the extract operation.
+ */
+RTEMS_INLINE_ROUTINE void _RBTree_Extract_unprotected(
+    RBTree_Control *the_rbtree,
+    RBTree_Node *the_node
+    )
+{
+  RBTree_Node *leaf, *target;
+  RBTree_Color victim_color;
+  RBTree_Direction dir;
+
+  if(!the_node) return;
+
+  /* check if min needs to be updated */
+  if (the_node == the_rbtree->first[RBT_LEFT]) {
+    if (the_node->child[RBT_RIGHT])
+      the_rbtree->first[RBT_LEFT] = the_node->child[RBT_RIGHT];
+    else {
+      the_rbtree->first[RBT_LEFT] = the_node->parent;
+      if(_RBTree_Are_nodes_equal((RBTree_Node *)the_rbtree, 
+            the_rbtree->first[RBT_LEFT]))
+        the_rbtree->first[RBT_LEFT] = NULL;
+    }
+  }
+  /* check if max needs to be updated: note, min can equal max (1 element) */
+  if (the_node == the_rbtree->first[RBT_RIGHT]) {
+    if (the_node->child[RBT_LEFT])
+      the_rbtree->first[RBT_RIGHT] = the_node->child[RBT_LEFT];
+    else {
+      the_rbtree->first[RBT_RIGHT] = the_node->parent;
+      if(_RBTree_Are_nodes_equal((RBTree_Node *)the_rbtree, 
+            the_rbtree->first[RBT_RIGHT]))
+        the_rbtree->first[RBT_RIGHT] = NULL;
+    }
+  }
+
+  /* if the_node has at most one non-null child then it is safe to proceed
+   * check if both children are non-null, if so then we must find a target node
+   * either max in node->child[RBT_LEFT] or min in node->child[RBT_RIGHT], 
+   * and replace the_node with the target node. This maintains the binary 
+   * search tree property, but may violate the red-black properties.
+   */
+
+  if (the_node->child[RBT_LEFT] && the_node->child[RBT_RIGHT]) {
+    target = the_node->child[RBT_LEFT]; /* find max in node->child[RBT_LEFT] */
+    while (target->child[RBT_RIGHT]) target = target->child[RBT_RIGHT];
+
+    /* if the target node has a child, need to move it up the tree into 
+     * target's position (target is the right child of target->parent) 
+     * when target vacates it. if there is no child, then target->parent 
+     * should become NULL. This may cause the coloring to be violated.
+     * For now we store the color of the node being deleted in victim_color.
+     */
+     leaf = target->child[RBT_LEFT];
+    if(leaf) { 
+      leaf->parent = target->parent;
+    } else {
+      /* fix the tree here if the child is a null leaf. */
+      _RBTree_Extract_validate_unprotected(target);
+    }
+    victim_color = target->color;
+    dir = target != target->parent->child[0];
+    target->parent->child[dir] = leaf;
+
+    /* now replace the_node with target */
+    dir = the_node != the_node->parent->child[0];
+    the_node->parent->child[dir] = target;
+
+    /* set target's new children to the original node's children */
+    target->child[RBT_RIGHT] = the_node->child[RBT_RIGHT];
+    the_node->child[RBT_RIGHT]->parent = target;
+    target->child[RBT_LEFT] = the_node->child[RBT_LEFT];
+    the_node->child[RBT_LEFT]->parent = target;
+
+    /* finally, update the parent node and recolor. target has completely 
+     * replaced the_node, and target's child has moved up the tree if needed.
+     * the_node is no longer part of the tree, although it has valid pointers
+     * still.
+     */
+    target->parent = the_node->parent;
+    target->color = the_node->color;
+  } else { 
+    /* the_node has at most 1 non-null child. Move the child in to 
+     * the_node's location in the tree. This may cause the coloring to be 
+     * violated. We will fix it later.
+     * For now we store the color of the node being deleted in victim_color.
+     */
+    leaf = the_node->child[RBT_LEFT] ? 
+              the_node->child[RBT_LEFT] : the_node->child[RBT_RIGHT];
+    if( leaf ) {
+      leaf->parent = the_node->parent;
+    } else {
+      /* fix the tree here if the child is a null leaf. */
+      _RBTree_Extract_validate_unprotected(the_node);
+    }
+    victim_color = the_node->color;
+
+    /* remove the_node from the tree */
+    dir = the_node != the_node->parent->child[0];
+    the_node->parent->child[dir] = leaf;
+  }
+
+  /* fix coloring. leaf has moved up the tree. The color of the deleted 
+   * node is in victim_color. There are three cases:
+   *   1. Deleted a red node, its child must be black. Nothing must be done. 
+   *   2. Deleted a black node and the child is red. Paint child black.
+   *   3. Deleted a black node and its child is black. This requires some
+   *      care and rotations.
+   */
+  if (victim_color == RBT_BLACK) { /* eliminate case 1 */
+    if (_RBTree_Is_red(leaf))
+      leaf->color = RBT_BLACK; /* case 2 */
+    else if(leaf)
+      _RBTree_Extract_validate_unprotected(leaf); /* case 3 */
+  }
+
+  /* Wipe the_node */
+  _RBTree_Set_off_rbtree(the_node);
+
+  /* set root to black, if it exists */
+  if (the_rbtree->root) the_rbtree->root->color = RBT_BLACK;
+}
+
+/** @brief Get the First Node (unprotected)
+ *
+ *  This function removes the minimum or maximum node from the_rbtree and 
+ *  returns a pointer to that node.  It does NOT disable interrupts to ensure
+ *  the atomicity of the get operation.
+ *
+ *  @param[in] the_rbtree is the rbtree to attempt to get the min node from.
+ *  @param[in] dir specifies whether to get minimum (0) or maximum (1)
+ *
+ *  @return This method returns the min or max node on the rbtree, or NULL.
+ *
+ *  @note This routine may return NULL if the RBTree is empty.
+ */
+RTEMS_INLINE_ROUTINE RBTree_Node *_RBTree_Get_unprotected(
+    RBTree_Control *the_rbtree,
+    RBTree_Direction dir
+    )
+{
+  RBTree_Node *the_node = the_rbtree->first[dir];
+  _RBTree_Extract_unprotected(the_rbtree, the_node);
+  return the_node;
+}
+
 #endif
 /* end of include file */
diff -upr -X dontdiff rtems/cpukit/score/inline/rtems/score/readyq.inl rtems-hwsched2/cpukit/score/inline/rtems/score/readyq.inl
--- rtems/cpukit/score/inline/rtems/score/readyq.inl	2010-07-28 17:10:58.000000000 -0400
+++ rtems-hwsched2/cpukit/score/inline/rtems/score/readyq.inl	2010-10-09 18:20:29.000000000 -0400
@@ -44,7 +44,11 @@ RTEMS_INLINE_ROUTINE void _Ready_queue_E
   Thread_Control             *the_thread
 )
 {
+  MAGIC_HWSCHED_START;
+  MAGIC_HWDS1_ENQUEUE;
   the_ready_queue->rq_ops.enqueue( the_ready_queue, the_thread );
+  MAGIC_HWDS1_ENQUEUE_END;
+  MAGIC_HWSCHED_END;
 }
 
 /*
@@ -61,7 +65,11 @@ RTEMS_INLINE_ROUTINE void _Ready_queue_E
   Thread_Control                   *the_thread
 )
 {
+  MAGIC_HWSCHED_START;
+  MAGIC_HWDS1_ENQUEUE_FIRST;
   the_ready_queue->rq_ops.enqueue_first( the_ready_queue, the_thread );
+  MAGIC_HWDS1_ENQUEUE_FIRST_END;
+  MAGIC_HWSCHED_END;
 }
 
 /** @brief  Ready queue Extract
@@ -73,7 +81,11 @@ RTEMS_INLINE_ROUTINE void _Ready_queue_E
   Thread_Control       *the_thread
 )
 {
+  MAGIC_HWSCHED_START;
+  MAGIC_HWDS1_EXTRACT;
   the_ready_queue->rq_ops.extract( the_ready_queue, the_thread );
+  MAGIC_HWDS1_EXTRACT_END;
+  MAGIC_HWSCHED_END;
 }
 
 
@@ -87,7 +99,13 @@ RTEMS_INLINE_ROUTINE Thread_Control *_Re
   Ready_queue_Control *the_ready_queue
 )
 {
- return ( the_ready_queue->rq_ops.first( the_ready_queue ) );
+  Thread_Control *t;
+  MAGIC_HWSCHED_START;
+ MAGIC_HWDS1_FIRST;
+ t= the_ready_queue->rq_ops.first( the_ready_queue );
+ MAGIC_HWDS1_FIRST_END;
+  MAGIC_HWSCHED_END;
+ return t;
 }
 
 /**
@@ -104,7 +122,9 @@ RTEMS_INLINE_ROUTINE void _Ready_queue_R
   Thread_Control       *the_thread
 )
 {
+  MAGIC_HWSCHED_START;
   the_ready_queue->rq_ops.requeue( the_ready_queue, the_thread );
+  MAGIC_HWSCHED_END;
 }
 
 /**@}*/
diff -upr -X dontdiff rtems/cpukit/score/inline/rtems/score/scheduler.inl rtems-hwsched2/cpukit/score/inline/rtems/score/scheduler.inl
--- rtems/cpukit/score/inline/rtems/score/scheduler.inl	2010-07-28 12:21:10.000000000 -0400
+++ rtems-hwsched2/cpukit/score/inline/rtems/score/scheduler.inl	2010-10-09 18:25:00.000000000 -0400
@@ -51,7 +51,9 @@ RTEMS_INLINE_ROUTINE void _Scheduler_Sch
     Scheduler_Control *the_scheduler 
 )
 {
+  MAGIC_HWSCHED_START;
   the_scheduler->s_ops.schedule( the_scheduler );
+  MAGIC_HWSCHED_END;
 }
 
 /**
@@ -76,7 +78,9 @@ RTEMS_INLINE_ROUTINE void _Scheduler_Blo
     Thread_Control *the_thread 
 )
 {
+  MAGIC_HWSCHED_START;
   the_scheduler->s_ops.block( the_scheduler, the_thread );
+  MAGIC_HWSCHED_END;
 }
 
 /**
@@ -89,7 +93,9 @@ RTEMS_INLINE_ROUTINE void _Scheduler_Unb
     Thread_Control *the_thread 
 )
 {
+  MAGIC_HWSCHED_START;
   the_scheduler->s_ops.unblock( the_scheduler, the_thread );
+  MAGIC_HWSCHED_END;
 }
 
 /**
@@ -100,7 +106,9 @@ RTEMS_INLINE_ROUTINE void* _Scheduler_Sc
   Thread_Control *the_thread
 )
 {
-  return ( the_scheduler->s_ops.sched_allocate( the_scheduler, the_thread ) );
+  void *v;
+  v= the_scheduler->s_ops.sched_allocate( the_scheduler, the_thread );
+  return v;
 }
 
 /**
@@ -111,7 +119,7 @@ RTEMS_INLINE_ROUTINE void _Scheduler_Sch
   Thread_Control *the_thread
 )
 {
-  return ( the_scheduler->s_ops.sched_free( the_scheduler, the_thread ) );
+  the_scheduler->s_ops.sched_free( the_scheduler, the_thread );
 }
 
 /**
diff -upr -X dontdiff rtems/cpukit/score/src/rbtreeextract.c rtems-hwsched2/cpukit/score/src/rbtreeextract.c
--- rtems/cpukit/score/src/rbtreeextract.c	2010-07-20 13:53:37.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/rbtreeextract.c	2010-10-09 19:09:23.000000000 -0400
@@ -17,207 +17,6 @@
 #include <rtems/score/rbtree.h>
 #include <rtems/score/isr.h>
 
-/** @brief  Validate and fix-up tree properties after deleting a node
- *
- *  This routine is called on a black node, @a the_node, after its deletion.
- *  This function maintains the properties of the red-black tree.
- *
- *  @note It does NOT disable interrupts to ensure the atomicity
- *        of the extract operation.
- */
-void _RBTree_Extract_validate_unprotected(
-    RBTree_Node *the_node
-    )
-{
-  RBTree_Node *parent, *sibling;
-  RBTree_Direction dir;
-
-  parent = the_node->parent; 
-  if(!parent->parent) return;
-
-  sibling = _RBTree_Sibling(the_node);
-
-  /* continue to correct tree as long as the_node is black and not the root */
-  while (!_RBTree_Is_red(the_node) && parent->parent) {
-
-    /* if sibling is red, switch parent (black) and sibling colors, 
-     * then rotate parent left, making the sibling be the_node's grandparent.
-     * Now the_node has a black sibling and red parent. After rotation, 
-     * update sibling pointer.
-     */
-    if (_RBTree_Is_red(sibling)) {
-      parent->color = RBT_RED;
-      sibling->color = RBT_BLACK;
-      dir = the_node != parent->child[0];
-      _RBTree_Rotate(parent, dir);
-      sibling = parent->child[!dir];
-    }
-
-    /* sibling is black, see if both of its children are also black. */
-    if (sibling && 
-        !_RBTree_Is_red(sibling->child[RBT_RIGHT]) && 
-        !_RBTree_Is_red(sibling->child[RBT_LEFT])) {
-        sibling->color = RBT_RED;
-        if (_RBTree_Is_red(parent)) {
-          parent->color = RBT_BLACK;
-          break;
-        }
-        the_node = parent; /* done if parent is red */
-        parent = the_node->parent;
-        sibling = _RBTree_Sibling(the_node);
-    } else if(sibling) {
-      /* at least one of sibling's children is red. we now proceed in two 
-       * cases, either the_node is to the left or the right of the parent. 
-       * In both cases, first check if one of sibling's children is black, 
-       * and if so rotate in the proper direction and update sibling pointer.
-       * Then switch the sibling and parent colors, and rotate through parent.
-       */
-      dir = the_node != parent->child[0];
-      if (!_RBTree_Is_red(sibling->child[!dir])) {
-        sibling->color = RBT_RED;
-        sibling->child[dir]->color = RBT_BLACK;
-        _RBTree_Rotate(sibling, !dir);
-        sibling = parent->child[!dir];
-      }
-      sibling->color = parent->color;
-      parent->color = RBT_BLACK;
-      sibling->child[!dir]->color = RBT_BLACK;
-      _RBTree_Rotate(parent, dir);
-      break; /* done */
-    }
-  } /* while */
-  if(!the_node->parent->parent) the_node->color = RBT_BLACK;
-}
-
-/** @brief Extract a Node (unprotected)
- *
- *  This routine extracts (removes) @a the_node from @a the_rbtree.
- *
- *  @note It does NOT disable interrupts to ensure the atomicity
- *        of the extract operation.
- */
-void _RBTree_Extract_unprotected(
-    RBTree_Control *the_rbtree,
-    RBTree_Node *the_node
-    )
-{
-  RBTree_Node *leaf, *target;
-  RBTree_Color victim_color;
-  RBTree_Direction dir;
-
-  if(!the_node) return;
-
-  /* check if min needs to be updated */
-  if (the_node == the_rbtree->first[RBT_LEFT]) {
-    if (the_node->child[RBT_RIGHT])
-      the_rbtree->first[RBT_LEFT] = the_node->child[RBT_RIGHT];
-    else {
-      the_rbtree->first[RBT_LEFT] = the_node->parent;
-      if(_RBTree_Are_nodes_equal((RBTree_Node *)the_rbtree, 
-            the_rbtree->first[RBT_LEFT]))
-        the_rbtree->first[RBT_LEFT] = NULL;
-    }
-  }
-  /* check if max needs to be updated: note, min can equal max (1 element) */
-  if (the_node == the_rbtree->first[RBT_RIGHT]) {
-    if (the_node->child[RBT_LEFT])
-      the_rbtree->first[RBT_RIGHT] = the_node->child[RBT_LEFT];
-    else {
-      the_rbtree->first[RBT_RIGHT] = the_node->parent;
-      if(_RBTree_Are_nodes_equal((RBTree_Node *)the_rbtree, 
-            the_rbtree->first[RBT_RIGHT]))
-        the_rbtree->first[RBT_RIGHT] = NULL;
-    }
-  }
-
-  /* if the_node has at most one non-null child then it is safe to proceed
-   * check if both children are non-null, if so then we must find a target node
-   * either max in node->child[RBT_LEFT] or min in node->child[RBT_RIGHT], 
-   * and replace the_node with the target node. This maintains the binary 
-   * search tree property, but may violate the red-black properties.
-   */
-
-  if (the_node->child[RBT_LEFT] && the_node->child[RBT_RIGHT]) {
-    target = the_node->child[RBT_LEFT]; /* find max in node->child[RBT_LEFT] */
-    while (target->child[RBT_RIGHT]) target = target->child[RBT_RIGHT];
-
-    /* if the target node has a child, need to move it up the tree into 
-     * target's position (target is the right child of target->parent) 
-     * when target vacates it. if there is no child, then target->parent 
-     * should become NULL. This may cause the coloring to be violated.
-     * For now we store the color of the node being deleted in victim_color.
-     */
-     leaf = target->child[RBT_LEFT];
-    if(leaf) { 
-      leaf->parent = target->parent;
-    } else {
-      /* fix the tree here if the child is a null leaf. */
-      _RBTree_Extract_validate_unprotected(target);
-    }
-    victim_color = target->color;
-    dir = target != target->parent->child[0];
-    target->parent->child[dir] = leaf;
-
-    /* now replace the_node with target */
-    dir = the_node != the_node->parent->child[0];
-    the_node->parent->child[dir] = target;
-
-    /* set target's new children to the original node's children */
-    target->child[RBT_RIGHT] = the_node->child[RBT_RIGHT];
-    the_node->child[RBT_RIGHT]->parent = target;
-    target->child[RBT_LEFT] = the_node->child[RBT_LEFT];
-    the_node->child[RBT_LEFT]->parent = target;
-
-    /* finally, update the parent node and recolor. target has completely 
-     * replaced the_node, and target's child has moved up the tree if needed.
-     * the_node is no longer part of the tree, although it has valid pointers
-     * still.
-     */
-    target->parent = the_node->parent;
-    target->color = the_node->color;
-  } else { 
-    /* the_node has at most 1 non-null child. Move the child in to 
-     * the_node's location in the tree. This may cause the coloring to be 
-     * violated. We will fix it later.
-     * For now we store the color of the node being deleted in victim_color.
-     */
-    leaf = the_node->child[RBT_LEFT] ? 
-              the_node->child[RBT_LEFT] : the_node->child[RBT_RIGHT];
-    if( leaf ) {
-      leaf->parent = the_node->parent;
-    } else {
-      /* fix the tree here if the child is a null leaf. */
-      _RBTree_Extract_validate_unprotected(the_node);
-    }
-    victim_color = the_node->color;
-
-    /* remove the_node from the tree */
-    dir = the_node != the_node->parent->child[0];
-    the_node->parent->child[dir] = leaf;
-  }
-
-  /* fix coloring. leaf has moved up the tree. The color of the deleted 
-   * node is in victim_color. There are three cases:
-   *   1. Deleted a red node, its child must be black. Nothing must be done. 
-   *   2. Deleted a black node and the child is red. Paint child black.
-   *   3. Deleted a black node and its child is black. This requires some
-   *      care and rotations.
-   */
-  if (victim_color == RBT_BLACK) { /* eliminate case 1 */
-    if (_RBTree_Is_red(leaf))
-      leaf->color = RBT_BLACK; /* case 2 */
-    else if(leaf)
-      _RBTree_Extract_validate_unprotected(leaf); /* case 3 */
-  }
-
-  /* Wipe the_node */
-  _RBTree_Set_off_rbtree(the_node);
-
-  /* set root to black, if it exists */
-  if (the_rbtree->root) the_rbtree->root->color = RBT_BLACK;
-}
-
-
 /*
  *  _RBTree_Extract
  *
diff -upr -X dontdiff rtems/cpukit/score/src/rbtreeinsert.c rtems-hwsched2/cpukit/score/src/rbtreeinsert.c
--- rtems/cpukit/score/src/rbtreeinsert.c	2010-07-20 13:36:46.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/rbtreeinsert.c	2010-10-09 19:09:23.000000000 -0400
@@ -17,109 +17,6 @@
 #include <rtems/score/rbtree.h>
 #include <rtems/score/isr.h>
 
-/** @brief Validate and fix-up tree properties for a new insert/colored node
- *
- *  This routine checks and fixes the Red-Black Tree properties based on 
- *  @a the_node being just added to the tree.
- * 
- *  @note It does NOT disable interrupts to ensure the atomicity of the
- *        append operation.
- */
-void _RBTree_Validate_insert_unprotected(
-    RBTree_Node    *the_node
-    )
-{
-  RBTree_Node *u,*g;
-
-  /* note: the insert root case is handled already */
-  /* if the parent is black, nothing needs to be done
-   * otherwise may need to loop a few times */
-  while (_RBTree_Is_red(_RBTree_Parent(the_node))) {
-    u = _RBTree_Parent_sibling(the_node);
-    g = the_node->parent->parent;
-    
-    /* if uncle is red, repaint uncle/parent black and grandparent red */
-    if(_RBTree_Is_red(u)) {
-      the_node->parent->color = RBT_BLACK;
-      u->color = RBT_BLACK;
-      g->color = RBT_RED;
-      the_node = g;
-    } else { /* if uncle is black */
-      RBTree_Direction dir = the_node != the_node->parent->child[0];
-      RBTree_Direction pdir = the_node->parent != g->child[0];
-      
-      /* ensure node is on the same branch direction as parent */
-      if (dir != pdir) {
-        _RBTree_Rotate(the_node->parent, pdir);
-        the_node = the_node->child[pdir];
-      }
-      the_node->parent->color = RBT_BLACK;
-      g->color = RBT_RED;
-      
-      /* now rotate grandparent in the other branch direction (toward uncle) */
-      _RBTree_Rotate(g, (1-pdir));
-    }
-  }
-  if(!the_node->parent->parent) the_node->color = RBT_BLACK;
-}
-
-
-
-/** @brief Insert a Node (unprotected)
- *
- *  This routine inserts @a the_node on the Red-Black Tree @a the_rbtree.
- *
- *  @retval 0 Successfully inserted.
- *  @retval -1 NULL @a the_node.
- *  @retval RBTree_Node* if one with equal value to @a the_node->value exists 
- *          in @a the_rbtree.
- *
- *  @note It does NOT disable interrupts to ensure the atomicity
- *        of the extract operation.
- */
-RBTree_Node *_RBTree_Insert_unprotected(
-    RBTree_Control *the_rbtree,
-    RBTree_Node *the_node
-    )
-{
-  if(!the_node) return (RBTree_Node*)-1;
-
-  RBTree_Node *iter_node = the_rbtree->root;
-
-  if (!iter_node) { /* special case: first node inserted */
-    the_node->color = RBT_BLACK;
-    the_rbtree->root = the_node;
-    the_rbtree->first[0] = the_rbtree->first[1] = the_node;
-    the_node->parent = (RBTree_Node *) the_rbtree;
-    the_node->child[RBT_LEFT] = the_node->child[RBT_RIGHT] = NULL;
-  } else { 
-    /* typical binary search tree insert, descend tree to leaf and insert */
-    while (iter_node) {
-      if(the_node->value == iter_node->value) return(iter_node);
-      RBTree_Direction dir = the_node->value > iter_node->value;
-      if (!iter_node->child[dir]) {
-        the_node->child[RBT_LEFT] = the_node->child[RBT_RIGHT] = NULL;
-        the_node->color = RBT_RED;
-        iter_node->child[dir] = the_node;
-        the_node->parent = iter_node;
-        /* update min/max */
-        if (_RBTree_Is_first(the_rbtree, iter_node, dir)) {
-          the_rbtree->first[dir] = the_node;
-        }
-        break;
-      } else {
-        iter_node = iter_node->child[dir];
-      }
-
-    } /* while(iter_node) */
-    
-    /* verify red-black properties */
-    _RBTree_Validate_insert_unprotected(the_node);
-  }
-  return (RBTree_Node*)0;
-}
-
-
 /*
  *  _RBTree_Insert
  *
diff -upr -X dontdiff rtems/cpukit/score/src/readyqedfextract.c rtems-hwsched2/cpukit/score/src/readyqedfextract.c
--- rtems/cpukit/score/src/readyqedfextract.c	2010-10-09 17:37:54.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/readyqedfextract.c	2010-10-02 21:04:19.000000000 -0400
@@ -97,6 +97,7 @@ void _Ready_queue_edf_Extract(
     if (!tmp_node) {
       /* TODO: remove after debugging */
       printk("invalid thread\n");
+      MAGIC_BREAKPOINT;
       while(1);
     }
     tmp_sched = _RBTree_Container_of(
diff -upr -X dontdiff rtems/cpukit/score/src/readyqpriorityrequeue.c rtems-hwsched2/cpukit/score/src/readyqpriorityrequeue.c
--- rtems/cpukit/score/src/readyqpriorityrequeue.c	2010-07-28 12:51:01.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/readyqpriorityrequeue.c	2010-10-09 18:18:48.000000000 -0400
@@ -46,10 +46,14 @@ void _Ready_queue_priority_Requeue(
 )
 {
   if ( !_Chain_Has_only_one_node( the_thread->sched.priority->ready_chain ) ) {
+  MAGIC_HWDS1_EXTRACT;
     _Chain_Extract_unprotected( &the_thread->Object.Node );
+  MAGIC_HWDS1_EXTRACT_END;
 
+  MAGIC_HWDS1_ENQUEUE;
     _Chain_Append_unprotected( the_thread->sched.priority->ready_chain, 
       &the_thread->Object.Node );
+  MAGIC_HWDS1_ENQUEUE_END;
   }
 }
 
diff -upr -X dontdiff rtems/cpukit/score/src/scheduleredf.c rtems-hwsched2/cpukit/score/src/scheduleredf.c
--- rtems/cpukit/score/src/scheduleredf.c	2010-07-29 17:56:42.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/scheduleredf.c	2010-10-09 18:18:10.000000000 -0400
@@ -37,12 +37,16 @@ void _Scheduler_edf_Release_job(
   ISR_Level       level;
 
   _ISR_Disable( level );
-  
+ 
+  MAGIC_HWSCHED_START;
+
   /* if the thread is already in a ready state, need to dequeue it from
    * the ready queue and then re-enqueue it with the new deadline.
    */
   if ( _States_Is_ready( the_period->owner->current_state ) ) {
+    MAGIC_HWDS1_EXTRACT;
     _Ready_queue_edf_Extract(&_Scheduler.ready_queue, the_period->owner);
+    MAGIC_HWDS1_EXTRACT_END;
 
     /* stamp this job with its new deadline */
     the_period->owner->sched.edf->deadline.value 
@@ -51,7 +55,9 @@ void _Scheduler_edf_Release_job(
     the_period->owner->sched.edf->last_duplicate = 
       &the_period->owner->Object.Node;
 
+    MAGIC_HWDS1_ENQUEUE;
     _Ready_queue_edf_Enqueue(&_Scheduler.ready_queue, the_period->owner);
+    MAGIC_HWDS1_ENQUEUE_END;
     _Scheduler_Schedule(&_Scheduler);
   } else {
     the_period->owner->sched.edf->deadline.value 
@@ -64,6 +70,8 @@ void _Scheduler_edf_Release_job(
   if (the_period->owner != _Thread_Heir)
     _Thread_Dispatch_necessary = true;
 
+  MAGIC_HWSCHED_END;
+
   _ISR_Enable( level );
 }
 
diff -upr -X dontdiff rtems/cpukit/score/src/scheduleredfschedallocate.c rtems-hwsched2/cpukit/score/src/scheduleredfschedallocate.c
--- rtems/cpukit/score/src/scheduleredfschedallocate.c	2010-07-29 18:33:43.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/scheduleredfschedallocate.c	2010-10-09 19:09:23.000000000 -0400
@@ -57,7 +57,6 @@ void* _Scheduler_edf_Sched_allocate (
     the_thread->sched.edf->last_duplicate = &the_thread->Object.Node;
     the_thread->sched.edf->this_thread = the_thread;
     the_thread->sched.edf->deadline.value = 0;
-    the_thread->sched.edf->absolute_deadline = 0;
   }
 
   return sched;
diff -upr -X dontdiff rtems/cpukit/score/src/scheduleredfschedupdate.c rtems-hwsched2/cpukit/score/src/scheduleredfschedupdate.c
--- rtems/cpukit/score/src/scheduleredfschedupdate.c	2010-07-29 18:24:51.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/scheduleredfschedupdate.c	2010-10-09 19:09:23.000000000 -0400
@@ -50,5 +50,5 @@ void _Scheduler_edf_Sched_update (
 {
   Scheduler_edf_Per_thread *sched = the_thread->sched.edf;
 
-  sched->absolute_deadline = the_thread->real_priority;
+  sched->deadline.value = the_thread->real_priority;
 }
diff -upr -X dontdiff rtems/cpukit/score/src/schedulerqueueyield.c rtems-hwsched2/cpukit/score/src/schedulerqueueyield.c
--- rtems/cpukit/score/src/schedulerqueueyield.c	2010-07-28 12:32:36.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/schedulerqueueyield.c	2010-10-05 13:16:01.000000000 -0400
@@ -60,12 +60,14 @@ void _Scheduler_queue_Yield( Scheduler_C
 
   executing = _Thread_Executing;
   _ISR_Disable( level );
-      
+  
+      MAGIC_HWSCHED_START;
       _Ready_queue_Requeue(&the_scheduler->ready_queue, executing);
 
-      _ISR_Flash( level );
+ //     _ISR_Flash( level );
 
       _Scheduler_Schedule(the_scheduler);
+      MAGIC_HWSCHED_END;
     
     if ( !_Thread_Is_heir( executing ) )
       _Thread_Dispatch_necessary = true;
diff -upr -X dontdiff rtems/cpukit/score/src/threadsettransient.c rtems-hwsched2/cpukit/score/src/threadsettransient.c
--- rtems/cpukit/score/src/threadsettransient.c	2010-08-13 11:50:35.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/threadsettransient.c	2010-10-02 21:04:19.000000000 -0400
@@ -58,14 +58,14 @@ void _Thread_Set_transient(
   uint32_t              old_state;
   
   _ISR_Disable( level );
-
+MAGIC_HWSCHED_START;
   old_state = the_thread->current_state;
   the_thread->current_state = _States_Set( STATES_TRANSIENT, old_state );
 
   if ( _States_Is_ready( old_state ) ) {
     _Ready_queue_Extract(&_Scheduler.ready_queue, the_thread);
   }
-
+MAGIC_HWSCHED_END;
   _ISR_Enable( level );
 
 }
diff -upr -X dontdiff rtems/cpukit/score/src/threadtickletimeslice.c rtems-hwsched2/cpukit/score/src/threadtickletimeslice.c
--- rtems/cpukit/score/src/threadtickletimeslice.c	2010-07-08 13:33:02.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/threadtickletimeslice.c	2010-10-05 11:49:05.000000000 -0400
@@ -71,7 +71,7 @@ void _Thread_Tickle_timeslice( void )
   /*
    *  The cpu budget algorithm determines what happens next.
    */
-
+MAGIC_HWSCHED_PAUSE;
   switch ( executing->budget_algorithm ) {
     case THREAD_CPU_BUDGET_ALGORITHM_NONE:
       break;
@@ -102,4 +102,5 @@ void _Thread_Tickle_timeslice( void )
 	break;
     #endif
   }
+  MAGIC_HWSCHED_UNPAUSE;
 }
diff -upr -X dontdiff rtems/cpukit/score/src/watchdogadjust.c rtems-hwsched2/cpukit/score/src/watchdogadjust.c
--- rtems/cpukit/score/src/watchdogadjust.c	2010-07-08 13:33:02.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/watchdogadjust.c	2010-10-04 20:24:06.000000000 -0400
@@ -67,7 +67,7 @@ void _Watchdog_Adjust(
           } else {
             units -= _Watchdog_First( header )->delta_interval;
             _Watchdog_First( header )->delta_interval = 1;
-
+            
             _ISR_Enable( level );
 
             _Watchdog_Tickle( header );
diff -upr -X dontdiff rtems/cpukit/score/src/watchdoginsert.c rtems-hwsched2/cpukit/score/src/watchdoginsert.c
--- rtems/cpukit/score/src/watchdoginsert.c	2010-09-28 19:57:08.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/watchdoginsert.c	2010-10-09 18:27:21.000000000 -0400
@@ -52,6 +52,8 @@ void _Watchdog_Insert(
     _ISR_Enable( level );
     return;
   }
+MAGIC_HWSCHED_START;
+  MAGIC_HWDS2_ENQUEUE;
 
   the_watchdog->state = WATCHDOG_BEING_INSERTED;
   _Watchdog_Sync_count++;
@@ -96,7 +98,9 @@ restart:
       *  mechanism used here WAS redesigned to address this.
       */
 
+  MAGIC_HWDS2_PAUSE;
 //     _ISR_Flash( level );
+  MAGIC_HWDS2_UNPAUSE;
 
      if ( the_watchdog->state != WATCHDOG_BEING_INSERTED ) {
        goto exit_insert;
@@ -119,5 +123,7 @@ restart:
 exit_insert:
   _Watchdog_Sync_level = insert_isr_nest_level;
   _Watchdog_Sync_count--;
+  MAGIC_HWDS2_END;
+  MAGIC_HWSCHED_END;
   _ISR_Enable( level );
 }
diff -upr -X dontdiff rtems/cpukit/score/src/watchdogremove.c rtems-hwsched2/cpukit/score/src/watchdogremove.c
--- rtems/cpukit/score/src/watchdogremove.c	2010-09-23 21:23:34.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/watchdogremove.c	2010-10-09 18:27:34.000000000 -0400
@@ -37,6 +37,8 @@ Watchdog_States _Watchdog_Remove(
   Watchdog_Control *next_watchdog;
 
   _ISR_Disable( level );
+  MAGIC_HWSCHED_START;
+  MAGIC_HWDS2_EXTRACT;
   previous_state = the_watchdog->state;
   switch ( previous_state ) {
     case WATCHDOG_INACTIVE:
@@ -67,6 +69,8 @@ Watchdog_States _Watchdog_Remove(
       break;
   }
   the_watchdog->stop_time = _Watchdog_Ticks_since_boot;
+  MAGIC_HWDS2_END;
+  MAGIC_HWSCHED_END;
   _ISR_Enable( level );
   return( previous_state );
 }
diff -upr -X dontdiff rtems/cpukit/score/src/watchdogtickle.c rtems-hwsched2/cpukit/score/src/watchdogtickle.c
--- rtems/cpukit/score/src/watchdogtickle.c	2010-07-08 13:33:02.000000000 -0400
+++ rtems-hwsched2/cpukit/score/src/watchdogtickle.c	2010-10-05 11:48:59.000000000 -0400
@@ -119,7 +119,6 @@ void _Watchdog_Tickle(
      }
 
      _ISR_Disable( level );
-
      the_watchdog = _Watchdog_First( header );
    } while ( !_Chain_Is_empty( header ) &&
              (the_watchdog->delta_interval == 0) );
